compiler/hash-typecheck/src/new/passes/symbol_resolution.rs

```diff
@@ -120,8 +276,644 @@ impl<'tc> SymbolResolutionPass<'tc> {
+            AstArgGroup::ImplicitArgs(args) => make_args_from_ast_args!(args),
+        }
+    }
+
+    /// Make [`DefArgsId`] from a list of AST argument groups, using
+    /// `make_args_from_ast_arg_group` to make each argument group.
+    fn make_def_args_from_ast_arg_groups(
+        &self,
+        groups: &[AstArgGroup],
+        originating_params: DefParamsId,
+    ) -> DefArgsId {
+        self.param_ops().create_def_args(groups.iter().enumerate().map(|(index, group)| {
+            DefArgGroupData {
+                args: self.make_args_from_ast_arg_group(group),
+                param_group: (originating_params, index),
+            }
+        }))
+    }
+
+    /// Wrap a term in a function call, given a list of arguments as a list of
+    /// [`AstArgGroup`].
+    ///
+    /// Creates a new [`TermId`], which is a term that represents the
+    /// function call, and might be nested depending on `args`.
+    fn wrap_term_in_fn_call_from_ast_args(&self, subject: TermId, args: &[AstArgGroup]) -> TermId {
+        let mut current_subject = subject;
+        for arg_group in args {
+            let args = self.make_args_from_ast_arg_group(arg_group);
+            current_subject = self.new_term(Term::FnCall(FnCallTerm {
+                subject: current_subject,
+                args,
+                implicit: matches!(arg_group, AstArgGroup::ImplicitArgs(_)),
+            }));
+        }
+        current_subject
+    }
+
+    /// Apply the given list of AST arguments to the given definition
+    /// parameters, checking that the lengths match in the process.
+    ///
+    /// If successful, returns the [`DefArgsId`] that represents the arguments.
+    ///
+    /// Otherwise, returns an error.
+    fn apply_ast_args_to_def_params(
+        &self,
+        def_params: DefParamsId,
+        args: &[AstArgGroup],
+    ) -> TcResult<DefArgsId> {
+        // @@Todo: implicit args
+
+        // First ensure that the number of parameter and argument groups match.
+        let created_def_args = self.make_def_args_from_ast_arg_groups(args, def_params);
+        if def_params.len() != created_def_args.len() {
+            return Err(TcError::WrongDefArgLength {
+                def_params_id: def_params,
+                def_args_id: SomeDefArgsId::Args(created_def_args),
+            });
+        }
+
+        // Then ensure that the number of parameters and arguments in each group
+        // match.
+        let mut errors: Vec<TcError> = vec![];
+        for (param_group_index, arg_group_index) in
+            def_params.to_index_range().zip(created_def_args.to_index_range())
+        {
+            let def_param_group =
+                self.stores().def_params().get_element((def_params, param_group_index));
+            let def_arg_group =
+                self.stores().def_args().get_element((created_def_args, arg_group_index));
+
+            if def_param_group.params.len() != def_arg_group.args.len() {
```

Comment:
Add a note here about default params, as it looks like it's not dealing with it.